cpp/src/parquet/file_reader.cc

```diff
@@ -212,6 +237,21 @@ class SerializedFile : public ParquetFileReader::Contents {
     file_metadata_ = std::move(metadata);
   }
 
+  void PreBuffer(const std::vector<int>& row_groups,
+                 const std::vector<int>& column_indices,
+                 const ::arrow::io::CacheOptions& options) {
+    cached_source_ =
+        std::make_shared<arrow::io::internal::ReadRangeCache>(source_, options);
```

Comment:
Be careful: the cache is unbounded. Does it mean you're willing to let all these row groups survive in memory until the reader gets destroyed?

Reply:
We could add a method to clear/remove the cache, or the user might recreate the ParquetFileReader (as the cache is shared among all readers created from the file anyways).

Reply:
I guess my question is: if I'm reading one record batch at a time (in streaming fashion), shouldn't the cache be per-record batch?

Reply:
It depends: reading multiple row groups at once might cache chunks that span row groups (and in general you want more opportunities to coalesce reads if you are aiming for throughput). For more control over that, a consumer could make repeated calls to `FileReader->RowGroup` to get individual RowGroupReaders that would cache only a single RowGroup.

Reply:
The other option would be to more fully separate the I/O and allow callers to pass a cache for a given read, and have `PreBuffer` instead pre-warm a supplied cache for the given row groups/column indices. Then you would have full control over the cache's lifetime, at the expense of a lot more complicated code (both for the caller and internally).

Reply:
I don't understand: is a separate `ParquetFileReader::Contents` created for each row group?

Reply:
No, on the contrary, there's only one instance of `Contents`, and hence a single shared cache right now (between all reads of the same file). However, lots of instances of `RowGroupReader::Contents` get created (one per row group per column), so it's not easy to cache each row group separately.

Perhaps I'm missing the point: what you'd like is a way to stream record batches out of a file, and have it internally clean up memory for each row group once the data has been fully read, right? (And, not pre-buffer more than some fixed number of row groups ahead of the current one.)

Reply:
I don't want anything specifically, but I wonder what users of the API may want. If you're reading a 10GB file, in a streaming fashion, from S3, you probably don't want to keep the 10GB in memory, right?

Reply:
That's fair. In our case, even if a dataset is large, individual files are smaller, so it's fine to buffer an entire file and then discard it. But I agree that for other use cases, this is not ideal.

A caller who is very concerned about memory might instead choose to explicitly read only one row group at a time to limit memory usage. This is rather annoying, though. 

We could create a separate cache per row group, but this means we lose some performance as we can't coalesce reads across row groups anymore. However that might be a worthwhile tradeoff for large files. Correct me if I'm wrong, though, but even that doesn't help much without more refactoring, since reading is organized along columns.

Reply:
I'm afraid I don't understand you. How do you "read one row group at a time" if the cache is shared at the FileReader level? Do you mean creating a new Parquet file reader for each row group?

Reply:
> Correct me if I'm wrong, though, but even that doesn't help much without more refactoring, since reading is organized along columns.

That was my original point, which you said was irrelevant...

Reply:
To be clear, I don't think this concern should block the PR, but the docstring should warn about it.

Reply:
GitHub acting up, sorry for the delay...

> > Correct me if I'm wrong, though, but even that doesn't help much without more refactoring, since reading is organized along columns.
> 
> That was my original point, which you said was irrelevant...

Apologies, I remember you had made this point on JIRA or the mailing list and was trying to find where (I still can't find where :/) I admit my understanding of the internals is poor, and your comment helped me understand what to look for better.

> I'm afraid I don't understand you. How do you "read one row group at a time" if the cache is shared at the FileReader level? Do you mean creating a new Parquet file reader for each row group?

Sorry, I was thinking about creating RowGroupReaders from a FileReader, one row group at a time: https://arrow.apache.org/docs/cpp/api/formats.html#_CPPv4N7parquet5arrow10FileReader8RowGroupEi

Each time this is called, it internally recreates the cache and buffers a new set of ranges. (Hence the comment above about "unsafe".) So this would let you read a row group's worth of data at a time, at the cost of more complexity for the user.

> To be clear, I don't think this concern should block the PR, but the docstring should warn about it.

I'll update it.

Reply:
Let me check this patch locally with dataset because this can drastically affect memory usage in multi-thread settings. And I'd say that if this explode memory usage for stream readers, that would be a blocker until we find an acceptable solution.

Reply:
Agreed, memory usage is part of why in the original discussion, we mentioned the I/O concurrency manager component - which among other things would let an application trade off read performance and bandwidth (which I brought up again today on the mailing list).